src/App/Datums.cpp

```diff
@@ -21,42 +23,318 @@
  ***************************************************************************/
 
 #include "PreCompiled.h"
+#ifndef _PreComp_
+#include <string>
+#endif
+
+#include <App/Document.h>
+#include <Base/Exception.h>
+#include <Base/Placement.h>
 
 #include "Datums.h"
-#include "Document.h"
-#include "Origin.h"
 
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
 
 using namespace App;
 
 PROPERTY_SOURCE(App::DatumElement, App::GeoFeature)
 PROPERTY_SOURCE(App::Plane, App::DatumElement)
 PROPERTY_SOURCE(App::Line, App::DatumElement)
+PROPERTY_SOURCE(App::Point, App::DatumElement)
+PROPERTY_SOURCE(App::LocalCoordinateSystem, App::GeoFeature)
 
-DatumElement::DatumElement()
+DatumElement::DatumElement(bool hideRole)
 {
-    ADD_PROPERTY_TYPE(Role, (""), 0, App::Prop_ReadOnly, "Role of the feature in the Origin");
+    ADD_PROPERTY_TYPE(Role,
+                      (""),
+                      0,
+                      App::Prop_ReadOnly,
+                      "Role of the datum in the local coordinate system.");
 
-    // Set placement to read-only
-    Placement.setStatus(Property::Hidden, true);
+    // The role is hidden by default. It is visible only when the datum is in a LCS
+    if (hideRole) {
+        Role.setStatus(Property::Hidden, true);
+    }
 }
 
 DatumElement::~DatumElement() = default;
 
-Origin* DatumElement::getOrigin()
+bool DatumElement::getCameraAlignmentDirection(Base::Vector3d& direction, const char* subname) const
 {
-    App::Document* doc = getDocument();
-    auto origins = doc->getObjectsOfType(App::Origin::getClassTypeId());
+    Q_UNUSED(subname);
+    Placement.getValue().getRotation().multVec(Base::Vector3d(0., 0., 1.), direction);
+
+    return true;
+}
+
+App::LocalCoordinateSystem* DatumElement::getLCS()
+{
+    auto inList = getInList();
+    for (auto* obj : inList) {
+        auto* lcs = dynamic_cast<App::LocalCoordinateSystem*>(obj);
+        if (lcs) {
+            return lcs;
+        }
+    }
```

Comment:
Really `std::find_if`?

Reply:
So that would be : 
```
App::LocalCoordinateSystem* DatumElement::getLCS()
{
    auto inList = getInList();
    
    auto it = std::find_if(inList.begin(), inList.end(), [](auto* obj) {
        return dynamic_cast<App::LocalCoordinateSystem*>(obj) != nullptr;
    });

    return (it != inList.end()) ? dynamic_cast<App::LocalCoordinateSystem*>(*it) : nullptr;
}
```
?
I find that the syntax is more difficult to read than the current and would rather keep it like it is. But if you prefer this I can replace it.

Reply:
Explicit only in the name of the function 'find_if'. But honestly this syntax is just not explicit. If I come accross something like this, it takes me at least 3 times longer to understand the intent than a simple for loop. And even now I'm still confused because this algo is using iterator. So you pass a lambda that is returning a bool. Just looking at it I'm still confused.
And what is sure is that I cannot write this out of my head. So everytime I would need to write something like this I'd have to use chatgpt. And debugging such is also painful as I struggle to understand.

If the syntax was simpler like in python : 
```
def getLCS(self):
    in_list = self.getInList()
    return next((obj for obj in in_list if isinstance(obj, App.LocalCoordinateSystem)), None)
```
Then I'd agree 100%. But it's not the case, the cpp syntax is heavy and hard to understand. It relies on the use of lambda + iterator and in my example even '?' operator. Which makes it less accessible for non-professional developpers.

Reply:
I'm for `std::find_if` here as well. It looks _difficult_  for the reason people are not used to see it though code base. After all everything can be open coded, but then we could stick with C++11 at most.

Reply:
This (probably) comes from forum topic [New code tending to be naive](https://forum.freecad.org/viewtopic.php?t=88567)

Reply:
Yes I think so too. This typical 'it was better before' topic. As if the old code is better, this is simply not true. Many of the original code is really bad. If anything, the new code is on average better. Of course you can cherry pick one example and show a clever refactor but whats the point?

And we are forgeting that the pace of development before was close to 0. And I mean 0. A PR could rot in the queue for months, even years before being reviewed.

And even then, what is the important code quality? I don't think it is about syntax and the use of algorithm instead of for loops. It is about breaking encapsulation or not testing if something is null causing a crash. Not deleting something causing a memory leak. Or doing some hacky things etc. It is not about using std::transform when push_back is 10x easier.

Reply:
@PaddleStroke, you are welcome to participate in #17817, where an attempt is made to establish general rules.